Add guidance for indicating nullability on DTOs (#18278)
dotnet/docs: docs/csharp/nullable-migration-strategies.md

What it does mean is that you can't use `T?` in a generic class or method declaration without constraints. For example, <xref:System.Linq.Enumerable.FirstOrDefault%60%601(System.Collections.Generic.IEnumerable%7B%60%600%7D)?displayProperty=nameWithType> won't be changed to return `T?`. You can overcome this limitation by adding either the `struct` or `class` constraint. With either of those constraints, the compiler knows how to generate code for both `T` and `T?`.

You may want to restrict the types used for a generic type argument to be non-nullable types. You can do that by adding the `notnull` constraint on that type argument. When that constraint is applied, the type argument must not be a nullable type.

> ## Updating Data Transfer Objects (DTOs) to indicate nullability

Reviewer: I'd consider making this guidance a bit more generalized - it's relevant to any late-initialization property scenario, not just to DTOs. DTO is indeed the prominent case where this occurs, but isn't the only one.

Reviewer: Late-initialized properties sounds good to me. I'd definitely avoid record types here, and also constructor-less types (since there may be a constructor, just not setting a particular property).

I think it's good to mention DTOs as well (the title can be something like "Late-initialized properties, DTOs and nullability").

> Indicating the nullability of properties that are late-initialized (i.e. set after construction) may require special consideration to ensure that your class continues to correctly express the original design intent.

Reviewer 2: nit: we don't use the Latin abbreviations.

```suggestion
Indicating the nullability of properties that are late-initialized, meaning set after construction, may require special consideration to ensure that your class continues to correctly express the original design intent.
```

Types that contain late-initialized properties, such as Data Transfer Objects (DTOs), are often instantiated by an external library, like a database ORM (Object Relational Mapper), a deserializer, or some other component that automatically populates properties from another source.

Consider the following DTO class, prior to enabling nullable reference types, that represents a student:

```csharp
class Student
{
    [Required]
    public string FirstName { get; set; }

    [Required]
    public string LastName { get; set; }

    public string VehicleRegistration { get; set; }
}
```

> The design intent (indicated in this case by the ``Required`` attribute) suggests that in this system, the ``FirstName`` and ``LastName`` properties are **mandatory**, and therefore not null.
> 
> The ``VehicleRegistration`` property is **not mandatory**, so may be null.

Reviewer 2: the code-fencing should be single back-ticks:

```suggestion
The design intent (indicated in this case by the `Required` attribute) suggests that in this system, the `FirstName` and `LastName` properties are **mandatory**, and therefore not null.

The `VehicleRegistration` property is **not mandatory**, so may be null.
```

> When we enable nullable reference types, we want to indicate on our DTO which of the properties may be nullable, consistent with our original intent:

Reviewer 2: Throughout the docs site, we use 2nd person:

```suggestion
When you enable nullable reference types, you want to indicate on our DTO which of the properties may be nullable, consistent with your original intent:
```

```csharp
class Student
{
    [Required]
    public string FirstName { get; set; }

    [Required]
    public string LastName { get; set; }

    public string? VehicleRegistration { get; set; }
}
```

> For this DTO, the only property that may be null is ``VehicleRegistration``.
> 
> However, the compiler raises ``CS8618`` warnings for both ``FirstName`` and ``LastName``, indicating the non-nullable properties are uninitialized.

Reviewer 2: single back-tick again:

```suggestion
For this DTO, the only property that may be null is ``VehicleRegistration``.

However, the compiler raises `CS8618` warnings for both `FirstName` and `LastName`, indicating the non-nullable properties are uninitialized.
```

> There are three options available to you that resolve the compiler warnings, in a way that maintains the original intent. Any of these options are valid; you should choose the one that best suits your coding style and design requirements.

Reviewer 2: no comma:

```suggestion
There are three options available to you that resolve the compiler warnings in a way that maintains the original intent. Any of these options are valid; you should choose the one that best suits your coding style and design requirements.
```

    {
        FirstName = firstName;
        LastName = lastName;
    }

    [Required]
    public string FirstName { get; set; }

    [Required]
    public string LastName { get; set; }

    public string? VehicleRegistration { get; set; }
}
```

This approach only works if the library that you use to instantiate the class supports passing parameters in the constructor.

In addition, a library may support passing *some* properties in the constructor, but not all.
For example, EF Core supports [constructor binding](/ef/core/modeling/constructors) for normal column properties, but not navigation properties.

Check the documentation on the library that instantiates your class, to understand the extent to which it supports constructor binding.

### Property with nullable backing field

If constructor binding won't work for you, one way to deal with this problem is to have a non-nullable property with a nullable backing field:

```csharp
private string? _firstName;

[Required]
public string FirstName
{
    set => _firstName = value;
    get => _firstName
           ?? throw new InvalidOperationException("Uninitialized " + nameof(FirstName))
}
```

> In this scenario, if the ``FirstName`` property is accessed before it has been initialized, then we throw an ``InvalidOperationException``, because the API contract has been used incorrectly.

Reviewer 2: single back-tick:

```suggestion
In this scenario, if the `FirstName` property is accessed before it has been initialized, then we throw an `InvalidOperationException`, because the API contract has been used incorrectly.
```

> In this scenario, if the ``FirstName`` property is accessed before it has been initialized, then we throw an ``InvalidOperationException``, because the API contract has been used incorrectly.

Reviewer 2: first person => the code:

```suggestion
In this scenario, if the ``FirstName`` property is accessed before it has been initialized, then the code throws an ``InvalidOperationException``, because the API contract has been used incorrectly.
```

> You should consider that some libraries may have special considerations when using backing fields. For example, EF Core must be configured to use [backing fields](/ef/core/modeling/backing-field) correctly.

Reviewer: Small note: in many (most?) cases EF will function well without any special configuration around backing fields, so I'd change `must` to `may need to`

### Initialize the property to null

> As a terser alternative to using a nullable backing field, or if the library that instantiates your class is not compatible with that approach, you can simply initialize the property to null directly, with the help of the null-forgiving operator (``!``):

Reviewer 2: code-fencing:

```suggestion
As a terser alternative to using a nullable backing field, or if the library that instantiates your class is not compatible with that approach, you can simply initialize the property to `null` directly, with the help of the null-forgiving operator (`!`):
```

```csharp
[Required]
public string FirstName { get; set; } = null!;

[Required]
public string LastName { get; set; } = null!;

public string? VehicleRegistration { get; set; }
```

> You will never observe an actual null value at runtime except as a result of a programming bug (e.g. by accessing the property before it has been properly initialized).

Reviewer: ```suggestion
You will never observe an actual null value at runtime except as a result of a programming bug (i.e. by accessing the property before it has been properly initialized).
```
